src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -171,120 +189,71 @@ private static void getSourcesRecursively(EntityTypeSource source, Map<String, E
-    String quotedSourceAlias = "\"" + sourceAlias + "\"";
-    String valueGetter = column.getValueGetter();
-    String filterValueGetter = column.getFilterValueGetter();
-
-    valueGetter = valueGetter.replace(":sourceAlias", quotedSourceAlias);
-
-    if (filterValueGetter != null) {
-        filterValueGetter = filterValueGetter.replace(":sourceAlias", quotedSourceAlias);
-    }
-
-    column.valueGetter(valueGetter);
-    column.filterValueGetter(filterValueGetter);
-
-    if (column.getDataType() instanceof ObjectType objectType) {
-        injectSourceAliasForObjectType(objectType, sourceAlias);
-    }
-
-    if (column.getDataType() instanceof ArrayType arrayType) {
-        injectSourceAliasForArrayType(arrayType, sourceAlias);
-    }
-
-    return column;
+    });
+  return column;
 }
 
-  private static void injectSourceAliasForObjectType(ObjectType objectType, String sourceAlias) {
+  private static void injectSourceAliasForObjectType(ObjectType objectType, EntityType entityType, Map<String, String> renamedAliases, String sourceAlias) {
     List<NestedObjectProperty> convertedProperties = objectType.getProperties()
       .stream()
-      .map(nestedField -> injectSourceAlias(nestedField, sourceAlias))
+      .map(nestedField -> injectSourceAlias(nestedField, entityType, renamedAliases, sourceAlias))
       .toList();
     objectType.properties(convertedProperties);
   }
 
-  private static void injectSourceAliasForArrayType(ArrayType arrayType, String sourceAlias) {
+  private static void injectSourceAliasForArrayType(ArrayType arrayType, EntityType entityType, Map<String, String> renamedAliases, String sourceAlias) {
     if (arrayType.getItemDataType() instanceof ArrayType nestedArrayType) {
-      injectSourceAliasForArrayType(nestedArrayType, sourceAlias);
+      injectSourceAliasForArrayType(nestedArrayType, entityType, renamedAliases, sourceAlias);
     }
     else if (arrayType.getItemDataType() instanceof ObjectType objectType) {
-      injectSourceAliasForObjectType(objectType, sourceAlias);
+      injectSourceAliasForObjectType(objectType, entityType, renamedAliases, sourceAlias);
     }
   }
 
+  private Stream<EntityTypeColumn> copyColumns(EntityTypeSource sourceFromParent, EntityType originalEntityType, Map<String, String> renamedAliases) {
+    return originalEntityType.getColumns()
+      .stream()
+      .map(column -> {
+        EntityTypeColumn newColumn = copyColumn(column, originalEntityType);
+        // Only treat newColumn as idColumn if outer source specifies to do so
+        newColumn.isIdColumn(newColumn.getIsIdColumn() == null ? null : Boolean.TRUE.equals(newColumn.getIsIdColumn()) && (sourceFromParent == null || Boolean.TRUE.equals(sourceFromParent.getUseIdColumns())));
+        injectSourceAlias(newColumn, originalEntityType, renamedAliases, newColumn.getSourceAlias());
+        newColumn.setSourceAlias(null);
+        return newColumn;
+      });
```

Comment:
To make it a little more efficient, I opted to keep this as 1 big single `map()` step instead of breaking it into several smaller ones. In the version with smaller lambdas, each lambda ends up referring to at least 1 method parameter, which means that the JVM has to allocate an object for each lambda in order to create a closure. With just 1 big lambda, then we only get the 1 object.

Totally a micro-optimization, but it's a simple one that doesn't harm readability, so why not ¯\\\_(ツ)_/¯